src/chain/network_event.rs

```diff
@@ -90,31 +100,98 @@ pub enum NetworkEvent {
-        match *self {
-            NetworkEvent::SectionInfo(ref self_si) => self_si.proves_successor(their_si, proofs),
-            _ => false,
+impl AccumulatingEvent {
+    pub fn from_network_event(
+        event: NetworkEvent,
+    ) -> (AccumulatingEvent, Option<SectionInfoSigPayload>) {
+        (event.payload, event.signature)
+    }
+
+    pub fn into_network_event(self) -> NetworkEvent {
+        NetworkEvent {
+            payload: self,
+            signature: None,
+        }
+    }
+
+    pub fn into_network_event_with(self, signature: Option<SectionInfoSigPayload>) -> NetworkEvent {
+        NetworkEvent {
+            payload: self,
+            signature,
         }
     }
 
-    /// Returns the payload if this is a `SectionInfo` event.
     pub fn section_info(&self) -> Option<&SectionInfo> {
         match *self {
-            NetworkEvent::SectionInfo(ref self_si) => Some(self_si),
+            AccumulatingEvent::SectionInfo(ref self_si) => Some(self_si),
             _ => None,
         }
     }
+}
+
+impl Debug for AccumulatingEvent {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        match self {
+            AccumulatingEvent::AddElder(ref id, _) => write!(formatter, "AddElder({}, _)", id),
+            AccumulatingEvent::RemoveElder(ref id) => write!(formatter, "RemoveElder({})", id),
+            AccumulatingEvent::Online(ref payload) => write!(
+                formatter,
+                "Online(new:{}, old:{})",
+                payload.new_public_id, payload.old_public_id
+            ),
+            AccumulatingEvent::Offline(ref id) => write!(formatter, "Offline({})", id),
+            AccumulatingEvent::OurMerge => write!(formatter, "OurMerge"),
+            AccumulatingEvent::NeighbourMerge(ref digest) => {
+                write!(formatter, "NeighbourMerge({:.14?})", HexFmt(digest))
+            }
+            AccumulatingEvent::SectionInfo(ref sec_info) => {
+                write!(formatter, "SectionInfo({:?})", sec_info)
+            }
+            AccumulatingEvent::ExpectCandidate(ref vote) => {
+                write!(formatter, "ExpectCandidate({:?})", vote)
+            }
+            AccumulatingEvent::PurgeCandidate(ref id) => {
+                write!(formatter, "PurgeCandidate({})", id)
+            }
+            AccumulatingEvent::TheirKeyInfo(ref payload) => {
+                write!(formatter, "TheirKeyInfo({:?})", payload)
+            }
+            AccumulatingEvent::AckMessage(ref payload) => {
+                write!(formatter, "AckMessage({:?})", payload)
+            }
+            AccumulatingEvent::SendAckMessage(ref payload) => {
+                write!(formatter, "SendAckMessage({:?})", payload)
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
+pub struct NetworkEvent {
+    pub payload: AccumulatingEvent,
+    pub signature: Option<SectionInfoSigPayload>,
```

Comment:
If this is only used for the `SectionInfo` variant, why not make it a part of the variant? A separate type for the payload would be unnecessary, then.

I'm guessing it's because the signature wouldn't apply to the event as a whole, but just to the section info, am I right? But would it be that much of a problem?

Reply:
If you are looking at the first commit, this is what it did :-)
It then require having a different type for the key for accumulating event.

Reply:
Ah, I see. Right, this way might be a bit more ergonomic, then :+1: